qiskit/chemistry/algorithms/ground_state_solvers/oovqe.py

```diff
@@ -0,0 +1,660 @@
+
+    def _energy_evaluation_oo(self, parameters: np.ndarray) -> Union[float, List[float]]:
+        """ Evaluate energy at given parameters for the variational form and parameters for
+        given rotation of orbitals.
+        Args:
+            parameters: parameters for variational form and orbital rotations.
+        Returns:
+            energy of the hamiltonian of each parameter.
+        Raises:
+            AquaError: Instantiate OrbitalRotation class and provide it to the
+                       orbital_rotation keyword argument
+        """
+
+        # slice parameter lists
+        if self._iterative_oo:
+            parameters_var_form = self._fixed_wavefunction_params
+            parameters_orb_rot = parameters
+        else:
+            parameters_var_form = parameters[:self.var_form_num_parameters]
+            parameters_orb_rot = parameters[self.var_form_num_parameters:]
+
+        logger.info('Parameters of wavefunction are: \n%s', repr(parameters_var_form))
+        logger.info('Parameters of orbital rotation are: \n%s', repr(parameters_orb_rot))
+
+        # rotate the orbitals
+        if self._orbital_rotation is None:
+            raise AquaError('Instantiate OrbitalRotation class and provide it to the '
+                            'orbital_rotation keyword argument')
+
+        self._orbital_rotation.orbital_rotation_matrix(parameters_orb_rot)
+
+        # preserve original qmolecule and create a new one with rotated orbitals
+        self._qmolecule_rotated = copy.copy(self._qmolecule)
+        OOVQE._rotate_orbitals_in_qmolecule(self._qmolecule_rotated, self._orbital_rotation)
+
+        # construct the qubit operator
+        operator, aux_operators = self._transformation._do_transform(self._qmolecule_rotated)
+        if isinstance(operator, LegacyBaseOperator):
+            operator = operator.to_opflow()
+        self._vqe.operator = operator
+        self._vqe.aux_operators = aux_operators
+        logger.debug('Orbital rotation parameters of matrix U at evaluation %d returned'
+                     '\n %s', self._vqe._eval_count, repr(self._orbital_rotation.matrix_a))
+        self._vqe.var_form._num_parameters = self.var_form_num_parameters
+
+        # compute the energy on given state
+        mean_energy = self._vqe._energy_evaluation(parameters=parameters_var_form)
+
+        return mean_energy
+
+    def solve(self,
+              driver: BaseDriver,
+              aux_operators: Optional[Union[List[FermionicOperator],
+                                            List[BosonicOperator]]] = None) \
+            -> Union[ElectronicStructureResult, 'VibronicStructureResult']:
+
+        # algorithm requires to have driver to initialize correctly the variables,
+        # however here one is free to change the driver type
+        if driver is not None:
+            self._driver = driver
+        else:
+            raise AquaError('Please specify a driver as you instantiate OOVQE class.')
```

Comment:
Since a driver can be passed in the constructor should this not be an additional test but now on self._driver and if not None raise an error. Although it seems you should always pass a driver in the constructor, i.e. its not optional there, and here it can be overridden so it should never be None here.

Reply:
driver should only be passed to `solve(...)` not in constructor.

Reply:
For the case of oovqe we need the driver to be passed also to the constructor in order to initialise the calculation and p We pass the driver also to the solve(...) @Brogis1

Reply:
Exactly, I tried to avoid it but is necessary by this algorithm. There is no simple way of avoiding it for now.

Reply:
> Since a driver can be passed in the constructor should this not be an additional test but now on self._driver and if not None raise an error. Although it seems you should always pass a driver in the constructor, i.e. its not optional there, and here it can be overridden so it should never be None here.

Yes, well spotted, it can be removed. It remains from the time when I was trying not to use driver in the constructor of OOVQE.

Reply:
You can move all logic that depends on the driver to the `solve` function. You don't have to create all the things in init, just move them to the `solve`.

Reply:
will require a minimum eigensolver factory. let's move to a separate PR and merge this one asap and have OOVQE included tomorrow.

Reply:
Ok it is movable (just tried) but like this only a part of OOVQE essential parameters will be initialized. 'solve' will initialize the rest when called. If you are ok with this fine.

Reply:
Thanks, I am fine with that! 
I think that's how it should be, as you can then reuse it for different problems without having to adjust the algorithm itself.

Reply:
Great, moved the logic to the `solve` function. Adapted the tests, all works. For the factory, we can sort out together with @mrossinek @adekusar-drl  Work in progress in: https://github.com/pbark/qiskit-aqua/tree/oovqe